src/monodroid/jni/strings.hh

```diff
@@ -0,0 +1,731 @@
+	class string_base
+	{
+	protected:
+		static constexpr TChar NUL = '\0';
+		static constexpr TChar ZERO = '0';
+
+	public:
+		explicit string_base (size_t initial_size = 0)
+			: buffer (initial_size)
+		{
+			// Currently we care only about `char`, maybe in the future we'll add support for `wchar` (if needed)
+			static_assert (std::is_same_v<TChar, char>, "TChar must be an 8-bit character type");
+
+			clear ();
+		}
+
+		explicit string_base (const string_segment &token)
+			: string_base (token.initialized () ? token.length () : 0)
+		{
+			if (token.initialized ())
+				assign (token.start (), token.length ());
+		}
+
+		force_inline size_t index () const noexcept
+		{
+			return idx;
+		}
+
+		force_inline size_t length () const noexcept
+		{
+			return idx;
+		}
+
+		force_inline bool empty () const noexcept
+		{
+			return length () == 0;
+		}
+
+		force_inline void set_index (size_t new_index) noexcept
+		{
+			idx = new_index;
+		}
+
+		force_inline void clear () noexcept
+		{
+			set_index (0);
+			buffer.get ()[0] = NUL;
+		}
+
+		force_inline void terminate () noexcept
+		{
+			buffer.get ()[idx] = NUL;
+		}
+
+		force_inline string_base& append (const TChar* s, size_t length) noexcept
+		{
+			if (s == nullptr || length == 0)
+				return *this;
+
+			resize_if_needed (length);
+			if constexpr (BoundsCheck) {
+				ensure_enough_space (length);
+			}
+
+			std::memcpy (buffer.get () + idx, s, length);
+			idx += length;
+			buffer.get ()[idx] = NUL;
+
+			return *this;
+		}
+
+		string_base& append (const char *s) noexcept
+		{
+			if (s == nullptr)
+				return *this;
+
+			return append (s, strlen (s));
+		}
+
+		string_base& append (int16_t i) noexcept
+		{
+			resize_if_needed (MAX_INTEGER_DIGIT_COUNT_BASE10);
+			append_integer (buffer.get (), i);
+			return *this;
+		}
+
+		string_base& append (uint16_t i) noexcept
+		{
+			resize_if_needed (MAX_INTEGER_DIGIT_COUNT_BASE10);
+			append_integer (i);
+			return *this;
+		}
+
+		string_base& append (int32_t i) noexcept
+		{
+			resize_if_needed (MAX_INTEGER_DIGIT_COUNT_BASE10);
+			append_integer (i);
+			return *this;
+		}
+
+		string_base& append (uint32_t i) noexcept
+		{
+			resize_if_needed (MAX_INTEGER_DIGIT_COUNT_BASE10);
+			append_integer (i);
+			return *this;
+		}
+
+		string_base& append (int64_t i) noexcept
+		{
+			resize_if_needed (MAX_INTEGER_DIGIT_COUNT_BASE10);
+			append_integer (i);
+			return *this;
+		}
+
+		string_base& append (uint64_t i) noexcept
+		{
+			resize_if_needed (MAX_INTEGER_DIGIT_COUNT_BASE10);
+			append_integer (i);
+			return *this;
+		}
+
+		force_inline string_base& assign (const TChar* s, size_t length) noexcept
+		{
+			idx = 0;
+			return append (s, length);
+		}
+
+		string_base& assign (const char *s) noexcept
+		{
+			if (s == nullptr)
+				return *this;
+
+			return assign (s, strlen (s));
+		}
+
+		force_inline bool next_token (size_t start_index, const TChar separator, string_segment& token) const noexcept
```

Comment:
Does "tokenization" *really* belong in the same type?

Reply:
I think yes. It's an operation on the string, thus it belongs in the string type.